In [7]:
import pandas as pd
import numpy as np
import plotly.express as px

In [2]:
wpData = pd.read_csv('world-population-data.csv', na_values=' ')
wpData.head(5)

Rank        Country Population (2024) Yearly Change  Net Change  \
0     1          India     1,450,935,791        0.89 %  12,866,195   
1     2          China     1,419,321,278       -0.23 %  -3,263,655   
2     3  United States       345,426,571        0.57 %   1,949,236   
3     4      Indonesia       283,487,931        0.82 %   2,297,864   
4     5       Pakistan       251,269,164        1.52 %   3,764,669   

  Density (P/Km²) Land Area (Km²) Migrants (net)  Fert. Rate  Med. Age  \
0             488       2,973,190       -630,830         2.0        28   
1             151       9,388,211       -318,992         1.0        40   
2              38       9,147,420      1,286,132         1.6        38   
3             156       1,811,570        -38,469         2.1        30   
4             326         770,880     -1,401,173         3.5        20   

  Urban Pop % World Share  
0        37 %     17.78 %  
1        66 %     17.39 %  
2        82 %      4.23 %  
3        59 %      3.47 %  
4        34 %      3.08 %

In [3]:
wpData.duplicated().sum()

np.int64(0)

### Convert Dtypes to the proper types

In [4]:
wpData['Country'] = wpData['Country'].astype(str)
wpData['Population (2024)'] = wpData['Population (2024)'].astype(str).str.replace(',', '').astype(int)
wpData['Yearly Change'] = wpData['Yearly Change'].astype(str).str.replace('%', '').astype(float)
wpData['Net Change'] = wpData['Net Change'].astype(str).str.replace(',', '').astype(int)
wpData['Density (P/Km²)'] = wpData['Density (P/Km²)'].astype(str).str.replace(',', '').astype(int)
wpData['Land Area (Km²)'] = wpData['Land Area (Km²)'].astype(str).str.replace(',', '').astype(int)
wpData['Migrants (net)'] = wpData['Migrants (net)'].astype(str).str.replace(',', '').astype(int)
wpData['Urban Pop %'] = wpData['Urban Pop %'].astype(str).str.replace('%', '').apply(pd.to_numeric, errors='coerce')
wpData['World Share'] = wpData['World Share'].astype(str).str.replace('%', '').astype(float)
wpData.info()
wpData.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rank               234 non-null    int64  
 1   Country            234 non-null    object 
 2   Population (2024)  234 non-null    int64  
 3   Yearly Change      234 non-null    float64
 4   Net Change         234 non-null    int64  
 5   Density (P/Km²)    234 non-null    int64  
 6   Land Area (Km²)    234 non-null    int64  
 7   Migrants (net)     234 non-null    int64  
 8   Fert. Rate         234 non-null    float64
 9   Med. Age           234 non-null    int64  
 10  Urban Pop %        216 non-null    float64
 11  World Share        234 non-null    float64
dtypes: float64(4), int64(7), object(1)
memory usage: 22.1+ KB


Rank        Country  Population (2024)  Yearly Change  Net Change  \
0     1          India         1450935791           0.89    12866195   
1     2          China         1419321278          -0.23    -3263655   
2     3  United States          345426571           0.57     1949236   
3     4      Indonesia          283487931           0.82     2297864   
4     5       Pakistan          251269164           1.52     3764669   

   Density (P/Km²)  Land Area (Km²)  Migrants (net)  Fert. Rate  Med. Age  \
0              488          2973190         -630830         2.0        28   
1              151          9388211         -318992         1.0        40   
2               38          9147420         1286132         1.6        38   
3              156          1811570          -38469         2.1        30   
4              326           770880        -1401173         3.5        20   

   Urban Pop %  World Share  
0         37.0        17.78  
1         66.0        17.39  
2         82.0         4.23  
3         59.0         3.47  
4         34.0         3.08

## Which country had the lowest yearly change in population? Saint Martin with -5.04% with a change of -1386 people

In [9]:
count = 8

all_df = wpData.sort_values('Yearly Change', ascending=False)
pos_df = all_df.head(count)
neg_df = all_df.tail(count)

fig, ax = plt.subplots()

pos_bars = ax.barh(pos_df['Country'], pos_df['Yearly Change'], color='#b8e0d4')
neg_bars = ax.barh(neg_df['Country'], neg_df['Yearly Change'], color='#fb6f92')

ax.bar_label(pos_bars, labels=pos_df['Net Change'], label_type='center')
ax.bar_label(neg_bars, labels=neg_df['Net Change'], label_type='center')

ax.invert_yaxis()

NameError: name 'plt' is not defined

### Which country had the highest yearly change in population? Chad with 5.07% with a change of 980,059 people

In [ ]:
wpData.sort_values('Yearly Change', ascending=False).head(5)

### Which country has the highest population density per square kilometer? Monaco with 25,927 people per square km

In [ ]:
wpData.sort_values('Density (P/Km²)', ascending=False).head(5)

### Which country has the highest level of immigration? USA with 1,286,132 migrants

In [ ]:
wpData.sort_values('Migrants (net)', ascending=False).head(5)



### Which country has the lowest level of emigration? Pakistan with -1,401,173 

In [ ]:
wpData.sort_values('Migrants (net)', ascending=True).head(5)